### PR TITLE
Remove warning supress

### PR DESCRIPTION
**Pull request by vhvb1989**
fixes: https://github.com/Azure/azure-sdk-for-cpp/issues/2474

Updating the code to remove the include of libcurl header from files not utilizing it.
The header is now included from the private_connection which is consumed by curl.cpp implementation.

Also adding https://github.com/nlohmann/json/pull/2673 to the json lib implementation.

With this PR, we can use MSVC on Windows again to fully compile with /analyse ON

### CODE REVIEW

File: sdk/core/azure-core/inc/azure/core/internal/json/json.hpp [hunk 1]
```diff
@@ -2176,7 +2176,7 @@ _az_JSON_HEDLEY_DIAGNOSTIC_PUSH
     class exception : public std::exception {
     public:
       /// returns the explanatory string
-      _az_JSON_HEDLEY_RETURNS_NON_NULL const char* what() const noexcept override
+      const char* what() const noexcept override
```
Comment (antkmsft) [inline sdk/core/azure-core/inc/azure/core/internal/json/json.hpp:2179]: Does this line needs to be modified?

**Review by antkmsft — APPROVED** [review]

File: sdk/core/azure-core/inc/azure/core/internal/json/json.hpp (hunk 1 above)
Comment (antkmsft) [reply sdk/core/azure-core/inc/azure/core/internal/json/json.hpp:2179]: I see, went to the link you possted.

Comment (vhvb1989) [reply sdk/core/azure-core/inc/azure/core/internal/json/json.hpp:2179]: just added a comment in the description.

This is from https://github.com/Azure/azure-sdk-for-cpp/pull/2677   which also affect us when using /analyze.  I am just bringing a patch from its repo

**Pull request merged**